pkg/s3select/json/reader.go

```diff
@@ -17,184 +17,52 @@
-			output, err = toSingleLineJSON(input, indexKey, value)
-			input = output
-			i++
-			return err == nil
-		})
-	default:
-		output, err = sjson.Set(input, currentKey, result.Value())
-	}
-
-	return output, err
-}
-
-type objectReader struct {
-	reader io.Reader
-	err    error
-
-	p     []byte
-	start int
-	end   int
-
-	escaped     bool
-	quoteOpened bool
-	curlyCount  uint64
-	endOfObject bool
-}
-
-func (or *objectReader) objectEndIndex(p []byte, length int) int {
-	for i := 0; i < length; i++ {
-		if p[i] == '\\' {
-			or.escaped = !or.escaped
-			continue
-		}
-
-		if p[i] == '"' && !or.escaped {
-			or.quoteOpened = !or.quoteOpened
-		}
-
-		or.escaped = false
-
-		switch p[i] {
-		case '{':
-			if !or.quoteOpened {
-				or.curlyCount++
-			}
-		case '}':
-			if or.quoteOpened || or.curlyCount == 0 {
-				break
-			}
-
-			if or.curlyCount--; or.curlyCount == 0 {
-				return i + 1
-			}
-		}
-	}
-
-	return -1
-}
-
-func (or *objectReader) Read(p []byte) (n int, err error) {
-	if or.endOfObject {
-		return 0, io.EOF
-	}
-
-	if or.p != nil {
-		n = copy(p, or.p[or.start:or.end])
-		or.start += n
-		if or.start == or.end {
-			// made full copy.
-			or.p = nil
-			or.start = 0
-			or.end = 0
-		}
-	} else {
-		if or.err != nil {
-			return 0, or.err
-		}
-
-		n, err = or.reader.Read(p)
-		or.err = err
-		switch err {
-		case nil:
-		case io.EOF, io.ErrUnexpectedEOF, io.ErrClosedPipe:
-			or.err = io.EOF
-		default:
-			return 0, err
-		}
-	}
-
-	index := or.objectEndIndex(p, n)
-	if index == -1 || index == n {
-		return n, nil
-	}
-
-	or.endOfObject = true
-	if or.p == nil {
-		or.p = p
-		or.start = index
-		or.end = n
-	} else {
-		or.start -= index
-	}
-
-	return index, nil
-}
-
-func (or *objectReader) Reset() error {
-	or.endOfObject = false
-
-	if or.p != nil {
-		return nil
-	}
-
-	return or.err
-}
-
 // Reader - JSON record reader for S3Select.
 type Reader struct {
-	args         *ReaderArgs
-	objectReader *objectReader
-	readCloser   io.ReadCloser
+	args       *ReaderArgs
+	decoder    *jstream.Decoder
+	valueCh    chan *jstream.MetaValue
+	readCloser io.ReadCloser
 }
 
 // Read - reads single record.
 func (r *Reader) Read() (sql.Record, error) {
-	if err := r.objectReader.Reset(); err != nil {
-		return nil, err
-	}
-
-	data, err := ioutil.ReadAll(r.objectReader)
-	if err != nil {
-		return nil, errJSONParsingError(err)
-	}
-
-	data = bytes.TrimSpace(data)
-	if len(data) == 0 {
+	v, ok := <-r.valueCh
+	if !ok {
+		if err := r.decoder.Err(); err != nil {
+			return nil, errJSONParsingError(err)
+		}
 		return nil, io.EOF
 	}
 
-	if !gjson.ValidBytes(data) {
-		return nil, errJSONParsingError(err)
-	}
+	var data []byte
+	var err error
 
-	if bytes.Count(data, []byte("\n")) > 0 {
-		var s string
-		if s, err = toSingleLineJSON("", "", gjson.ParseBytes(data)); err != nil {
-			return nil, errJSONParsingError(err)
+	if v.ValueType == jstream.Object {
+		kvs := v.Value.(jstream.KVS)
+		if len(kvs) == 0 {
```

Comment:
Empty object is also handled in jstream!

Reply:
I know neat! - changed